gadgets/src/permutation.rs

```diff
@@ -0,0 +1,312 @@
+    plonk::{Advice, Column, ConstraintSystem, Error, Expression, Selector},
+    poly::Rotation,
+};
+use itertools::Itertools;
+
+/// Config for PermutationChipConfig
+#[derive(Clone, Debug)]
+pub struct PermutationChipConfig<F> {
+    // column
+    acc_fingerprints: Column<Advice>,
+    row_fingerprints: Column<Advice>,
+    alpha: Column<Advice>,
+    power_of_gamma: Vec<Column<Advice>>,
+    // selector
+    q_row_non_first: Selector, // 1 between (first, end], exclude first
+    q_row_enable: Selector,    // 1 for all rows (including first)
+
+    _phantom: PhantomData<F>,
+}
+
+type PermutationAssignedCells<F> = (
+    AssignedCell<F, F>,
+    AssignedCell<F, F>,
+    AssignedCell<F, F>,
+    AssignedCell<F, F>,
+);
+
+impl<F: Field> PermutationChipConfig<F> {
+    /// assign
+    pub fn assign(
+        &self,
+        region: &mut Region<'_, F>,
+        alpha: Value<F>,
+        gamma: Value<F>,
+        acc_fingerprints_prev: Value<F>,
+        col_values: &[Vec<Value<F>>],
+        prefix: &'static str,
+    ) -> Result<PermutationAssignedCells<F>, Error> {
+        self.annotate_columns_in_region(region, prefix);
+
+        // get accumulated fingerprints of each row
+        let fingerprints =
+            get_permutation_fingerprints(col_values, alpha, gamma, acc_fingerprints_prev);
+
+        let power_of_gamma = {
+            let num_of_col = col_values.get(0).map(|row| row.len()).unwrap_or_default();
+            std::iter::successors(Some(Value::known(F::ONE)), |prev| (*prev * gamma).into())
+                .take(num_of_col)
+                .collect::<Vec<Value<F>>>()
+        };
+
+        let mut last_fingerprint_cell = None;
+        let mut alpha_first_cell = None;
+        let mut gamma_first_cell = None;
+        let mut acc_fingerprints_prev_cell = None;
+        for (offset, (row_acc_fingerprints, row_fingerprints)) in fingerprints.iter().enumerate() {
+            // skip first fingerprint for its prev_fingerprint
+            if offset != 0 {
+                self.q_row_non_first.enable(region, offset)?;
+            }
+
+            self.q_row_enable.enable(region, offset)?;
+
+            let row_acc_fingerprint_cell = region.assign_advice(
+                || format!("acc_fingerprints at index {}", offset),
+                self.acc_fingerprints,
+                offset,
+                || *row_acc_fingerprints,
+            )?;
+
+            region.assign_advice(
+                || format!("row_fingerprints at index {}", offset),
+                self.row_fingerprints,
+                offset,
+                || *row_fingerprints,
+            )?;
+
+            let alpha_cell = region.assign_advice(
+                || format!("alpha at index {}", offset),
+                self.alpha,
+                offset,
+                || alpha,
+            )?;
+            let gamma_cells = self
+                .power_of_gamma
+                .iter()
+                .zip(power_of_gamma.iter())
```

Comment:
I believe `power_of_gamma` first element is `gamma^0 = 1` so we should skip the first element here.

Reply:
Nice catch!
Fix in latest commit.
Previous this not trigger unittest error due to gamma is set as `1`
I also update hardcode challenge value to assure it can capture this bug, besides use zip_eq to be more error prune